src/lib/support/CHIPSparseArray.h

```diff
@@ -0,0 +1,193 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2016-2017 Nest Labs, Inc.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <stddef.h>
+
+#include <core/CHIPEncoding.h>
+#include <core/CHIPError.h>
+#include <support/Base64.h>
+#include <support/CHIPMem.h>
+#include <support/CodeUtils.h>
+#include <support/SafeInt.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+
+template <uint16_t Capacity>
+class SparseU64Array
+{
+
+public:
+    SparseU64Array() { CHIP_ZERO_AT(mArray); }
+    /**
+     * @brief
+     *   Serialize the sparse array into a base 64 encoded string.
+     *   Only the values till mNextAvailable index are encoded.
+     *   The empty indexes between 0, and mNextAvailable, are also
+     *   encoded.
+     *
+     * @param[in] buf Buffer where serialized string is written
+     * @param[in] buflen Length of buf
+     * @return pointer to buf, or nullptr in case of error
+     */
+    const char * Serialize(char * buf, uint16_t & buflen)
```

Comment:
Should we maybe serliaze as binary instead? Or name this as 'SerializeBase64'.

My general expectation of a format would be length-prefix array of only valid data. Why do we serialize 0 values as well? Some description about design choices on this class would be useful (seels like zero-copy and optimize for fast inserts, at the cost of not allowing 0 values).

Why is the array called 'sparse'? Is this more like a U64Set?

Reply:
Base64 encoding will be useful for persistent storage (where key, values are strings). I'll rename the function to reflect encoding.

As @rwalker-apple also commented above, we can try to squeeze out 0 values from it at serialization. But, I would like to compact the array instead of just using non zero values for encoding. This will help deserialize to reproduce the original array.

Sparse, since not all contiguous values are valid (some could be 0).

Reply:
I see here about the String Key/String Value. I strongly believe moving to a <size,  span of bytes> model is more efficient for constrained embedded devices, and will avoid many serialization/deserialization roundtrips involving strings.